File: backend/app/schemas/log.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel


class LogResponse(BaseModel):
    id: int
    level: str
    scope: str
    message: str
    domain_id: int | None
    run_id: int | None
    metadata_json: dict | None
    created_at: datetime

    @classmethod
    def from_model(cls, model: object) -> "LogResponse":
        return cls(
            id=model.id,
            level=model.level,
            scope=model.scope,
            message=model.message,
            domain_id=model.domain_id,
            run_id=model.run_id,
            metadata_json=model.metadata_json,
            created_at=model.created_at,
        )

File: frontend/src/types/log.ts
export interface LogRecord {
  id: number;
  level: string;
  scope: string;
  message: string;
  domain_id: number | null;
  run_id: number | null;
  metadata_json: {
    trigger_source?: string;
    final_provider?: string;
    provider_attempts?: Array<{
      provider: string;
      status: string;
      error_code?: string | null;
      error_message?: string | null;
    }>;
    error_code?: string | null;
    error_message?: string | null;
  } | null;
  created_at: string;
}
